emacs/archive/server/gnuserv.1
.TH GNUSERV 1 "" "Server / Client enhancements for GNU Emacs"
.UC 4
.SH NAME
gnuserv, gnuclient, gnudoit \- Server / Client Enhancements for GNU Emacs
.SH SYNOPSIS
.B gnuclient
[-h hostname] [-q] [-p port] [-r pathname] [[+line] path] ...
.br
.B gnudoit 
[-h hostname] [-q] [-p port] [sexpr] ...
.br
.B gnuserv
.SH DESCRIPTION
\fBgnuclient\fP allows the user to request a running GNU Emacs process to
edit the named files or directories. 
.PP
\fBgnudoit\fP allows the user to request a running GNU Emacs process to
evaluate the given arguments inside a progn lisp form.
.PP
\fBgnuserv\fP is the server program that is set running by GNU Emacs to
handle all incoming and outgoing requests. It is not usually invoked directly, but is
started from GNU Emacs by the lisp form (server-start).
.PP
The options are:
.P
.TP 10
\fB-q\fP
This option means that both gnuclient and gnudoit will exit once connection has
been made with the GNU Emacs process. Normally gnuclient waits until all of the
files on the command line have been finished with (their buffers killed) by the
GNU Emacs process. Normally gnudoit
waits around for evaluation of its arguments by the GNU Emacs process, and prints
the results or error conditions.
.TP 10
\fB-h\fP hostname
This option specifies the host machine which should be running gnuserv. If this option
is not specified then the value of the environment variable
``GNU_HOST'' is used if set, otherwise the hostname of the machine running the
program is used.
Note that an internet address may be specified instead of a 
hostname which can speed up connections to the server by quite a factor,
especially if the client machine is running YP.
.TP 10
\fB-p\fP port
The service is called ``gnuserv'' in the services database, but a
different service port may be specified with this option.
The server must also be using the same service port. If this option is not
specified, but the value of the environment variable ``GNU_PORT'' is
specified, then it is used instead. Note that gnuserv doesn't use the command
line option since it is usually started from within GNU Emacs.
.TP 10
\fB-p\fP pathname
The pathname argument is needed such that the server knows how to specify the
root directory of a remote machine. \fBgnuclient\fP prepends this string to each
path argument given. For example, if you were trying to edit a file on a client
machine called otter, whose root directory
was accessible \fIfrom\fP the server machine via the path /net/otter, then this
argument should be set to '/net/otter'. If this option is omitted,
then the value is taken from the environment variable ``GNU_NODE'', if set, or
the empty string otherwise.
.TP 10
\fBpath\fP
This is the path of the file to be edited. If the file is a directory, then the
directory browser dired is usually invoked instead.
.TP 10
\fBsexpr\fP
This is part of a gnu emacs LISP expression to evaluate. All the arguments are
bundled together and wrapped in a progn form.
.PP
.SH SETUP
In order to use the programs, the file gnuserv.el can be copied into a
directory on your GNU Emacs load-path, and loaded into GNU Emacs by the lisp
form (load "gnuserv"). The server can then be initialized by the GNU Emacs
lisp form (server-start).
.PP
Additionally, the patch to src/x11fns.c can be applied to allow GNU Emacs to
de-iconify and raise its edit window upon request under certain X11 window
managers.
.SH EXAMPLE
.PP
.TP 10
gnudoit -q '(server-make-window-visible)' '(mh-smail)'
.TP 10
gnuclient -h otter -p /net/otter /tmp/*
.SH SECURITY
\fBgnuserv\fP admits a limited form of security at the machine level. By default only
connections from the host where the server is running will be allowed. All other
server connections will be rejected with a cryptic message (which is displayed
only by gnudoit). Alternatively, if the
variable ``GNU_SECURE'' can be found in gnuserv's environment, and it names a
readable filename, then this file is opened and assumed to be a list of
hosts, one per line, from which the server will allow requests. Note that a
host may be either a internet address, or a hostname. If this file contains a
lot of hostnames then the server may take quite a time to start up.
.SH KNOWN BUGS
.PP
If GNU Emacs attempts to send a string containing a newline character to
\fBgnuserv\fP, then \fBgnuserv\fP will die.
.SH FILES
.PP
.TP 8
.B /etc/services
.TP 8
.B ~/.emacs
\fIEmacs\fP customization file, see \fIemacs(1)\fP.
.SH AUTHOR.
Andy Norman (ange%anorman@hplabs.hp.com).


emacs/archive/server/gnuserv.c
/* -*-C-*-
 Server code for handling requests from clients and forwarding them
 on to the GNU Emacs process.

 This file is not part of GNU Emacs (yet).
 Copyright (c) 1989 Andrew P. Norman

 Copying is permitted under those conditions described by the GNU
 Emacs General Public License.

 Author: Andy Norman (ange%anorman@hplabs.hp.com)

 Please mail bugs and suggestions to the author at the above address.
*/

#include "gnulib.h"

struct entry {
  u_long host_addr;
  struct entry *next;
};

struct entry *permitted_hosts[TABLE_SIZE];


/*
  permitted -- return whether a given host is allowed to connect to the server.
*/
int permitted(host_addr)
     u_long host_addr;
{
  int key;
  struct entry *entry;
  
  /* First find the hash key */
  key = HASH(host_addr) % TABLE_SIZE;
  
  /* Now check the chain for that hash key */
  for(entry=permitted_hosts[key]; entry != NULL; entry=entry->next)
    if (host_addr == entry->host_addr) 
      return(TRUE);

  return(FALSE);

} /* permitted */


/* 
  add_host -- add the given host to the list of permitted hosts, provided it isn't
              already there.
*/	
void add_host(host_addr)
     u_long host_addr;
{
  int key;
  struct entry *new_entry;
  
  if (!permitted(host_addr)) {
    if ((new_entry = (struct entry *) malloc(sizeof(struct entry))) == NULL) {
      fprintf(stderr,"%s: unable to malloc space for permitted host entry\n",
	      progname);
      exit(1);
    }; /* if */

    new_entry->host_addr = host_addr;
    key = HASH(host_addr) % TABLE_SIZE;
    new_entry->next = permitted_hosts[key];
    permitted_hosts[key] = new_entry;
  }; /* if */

} /* add_host */


void setup_table()
{
  FILE *host_file;
  char *file_name;
  char hostname[HOSTNAMSZ];
  u_long host_addr;
  int i;
  
  /* Make sure every entry is null */
  for (i=0; i<TABLE_SIZE; i++)
    permitted_hosts[i] = NULL;

  if (((file_name = getenv("GNU_SECURE")) != NULL &&
       (host_file = fopen(file_name,"r")) != NULL)) {
    while ((fscanf(host_file,"%s",hostname) != EOF)) 
      if ((host_addr = internet_addr(hostname)) != -1)
	add_host(host_addr);

    fclose(host_file);
  }
  else { /* file not available from environment so only allow this host */
    gethostname(hostname,HOSTNAMSZ);

    if ((host_addr = internet_addr(hostname)) == -1) {
      fprintf(stderr,"%s: unable to find %s in /etc/hosts or from YP", 
	      progname,hostname);
      exit(1);
    }; /* if */

    add_host(host_addr);
  }; /* else */

} /* setup_table */


/*
  init -- initialize server, returning a socket that can be listened on.
*/
int init()
{
  int ls;			/* socket descriptor */
  int chan;			/* unix channel number */
  struct servent *sp;		/* pointer to service information */
  struct sockaddr_in myaddr_in;	/* for local socket address */
  char *ptr;			/* ptr for getenv */

  setup_table();

  for(chan=3; chan < _NFILE; close(chan++)) /* close unwanted channels */
    ;

  /* clear out address structure */
  memset ((char *)&myaddr_in, 0, sizeof(struct sockaddr_in));
  
  /* Set up address structure for the listen socket. */
  myaddr_in.sin_family = AF_INET;
  myaddr_in.sin_addr.s_addr = INADDR_ANY;
  
  /* Find the information for the gnu server
   * in order to get the needed port number.
   */
  if ((ptr=getenv("GNU_PORT")) != NULL)
    myaddr_in.sin_port = htons(atoi(ptr));
  else {
    if ((sp = getservbyname ("gnuserv", "tcp")) == NULL) {
      fprintf(stderr,"%s: unable to get \"gnuserv\" from /etc/services or from YP\n",
	      progname);
      exit(1);
    }; /* if */
    
    myaddr_in.sin_port = sp->s_port;
  }; /* else */
  
  /* Create the listen socket. */
  if ((ls = socket (AF_INET, SOCK_STREAM, 0)) == -1) {
    perror(progname);
    fprintf(stderr,"%s: unable to create socket\n",progname);
    exit(1);
  }; /* if */
  
  /* Bind the listen address to the socket. */
  if (bind(ls, &myaddr_in, sizeof(struct sockaddr_in)) == -1) {
    perror(progname);
    fprintf(stderr,"%s: unable to bind socket\n",progname);
    exit(1);
  }; /* if */

  /* Initiate the listen on the socket so remote users
   * can connect. 
   */
  if (listen(ls,20) == -1) {
    perror(progname);
    fprintf(stderr,"%s: unable to listen\n",progname);
    exit(1);
  }; /* if */
  
/*  setpgrp(); */
  return(ls);
  
} /* init */


/*
  handle_request -- accept a request from a client and send the information
                    to stdout (the gnu process).
*/
void handle_request(ls)
int ls;
{
  char buf[BUFSIZ];
  int s;
  int len;
  int addrlen = sizeof(struct sockaddr_in);
  struct sockaddr_in peeraddr_in;	/* for peer socket address */

  memset ((char *)&peeraddr_in,0,sizeof(struct sockaddr_in));

  if ((s = accept(ls,&peeraddr_in,&addrlen)) == -1) {
    perror(progname);
    fprintf(stderr,"%s: unable to accept\n",progname);
    exit(1);
  }; /* if */

  /* Check that access is allowed - if not return crud to the client */
  if (!permitted(peeraddr_in.sin_addr.s_addr)) {
    send_string(s,"gnudoit: Connection refused\ngnudoit: unable to connect to remote");
    close(s);
    return;
  }; /* if */

  printf("%d ",s);
  
  while (len = recv(s,buf,BUFSIZ,0)) {
    if (len < 0) {
      perror(progname);
      fprintf(stderr,"%s: unable to recv\n",progname);
      exit(1);
    }; /* if */
	
    buf[len] = '\0';
    printf("%s",buf);
  }; /* while */
  
  printf("\n");

} /* handle_request */


/*
  handle_response -- accept a response from stdin (the gnu process) and pass the
                     information on to the relevant client.
*/
void handle_response()
{
  char buf[BUFSIZ];
  char response[BUFSIZ];
  int s;
  int len;

  if ((len = read(0,buf,BUFSIZ)) < 0) {
    perror(progname);
    fprintf(stderr,"%s: unable to read\n",progname);
    exit(1);
  }; /* if */
      
  /* parse the response from gnu */
  response[0] = '\0';
  sscanf(buf,"%d:%[^\n]\n", &s, response);
  
  /* Send a response back to the client. */
  send_string(s,response);
  close(s);

} /* handle_response */


main(argc,argv)
int argc;
char *argv[];
{
  int ls;

  progname = argv[0];
  ls = init();			/* init and get a socket to listen on */

  while (1) {
    int rmask = (1 << ls) + 1;
    
    if (select(ls + 1,&rmask,0,0,0) < 0) {
      perror(progname);
      fprintf(stderr,"%s: unable to select\n",progname);
      exit(1);
    }
    
    if (rmask & (1 << ls))	/* from socket (client process) */
      handle_request(ls);
    
    if (rmask & 1) 		/* from stdin (gnu process) */
      handle_response();

  }; /* while */

} /* main */

